### PR TITLE
Initial version with Spring Boot 1.3.X and below

### PR DESCRIPTION
**Pull request by lordofthejars**
Initial commit for adding support for creating Spring Boot runnable jars in ShrinkWrap. Currently only Spring Boot 1.3.X or prior are supported.

### CODE REVIEW

File: impl/src/main/java/org/shrinkwrap/springboot/impl/SpringBoot13ArchiveImpl.java [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+        for (JavaArchive l : launcher) {
+            this.merge(l);
+        }
+        return this;
+    }
+
+    /*
+     * (non-Javadoc)
+     *
+     * @see org.jboss.declarchive.impl.base.ContainerBase#getManifestPath()
```
Comment (eddumelendez) [inline impl/src/main/java/org/shrinkwrap/springboot/impl/SpringBoot13ArchiveImpl.java]: wrong package... it should be `org.jboss.shrinkwrap.impl.base.ContainerBase`

File: impl/src/main/java/org/shrinkwrap/springboot/impl/SpringBoot13ArchiveImpl.java [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+    @Override
+    protected ArchivePath getLibraryPath() {
+        return PATH_LIBRARY;
+    }
+
+    /*
+     * (non-Javadoc)
+     *
+     * @see org.jboss.declarchive.impl.base.ContainerBase#getResourcePath()
```
Comment (eddumelendez) [inline impl/src/main/java/org/shrinkwrap/springboot/impl/SpringBoot13ArchiveImpl.java]: wrong package... it should be `org.jboss.shrinkwrap.impl.base.ContainerBase`

**Comment by bartoszmajsak** [comment]
Ping @ALRubinger :)

**Comment by MatousJobanek** [comment]
What needs to be done to make the ftest running?
I run `mvn clean test` and I got:
```
-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Running org.shrinkwrap.springboot.app.SpringBoot13Test
Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 3.89 sec <<< FAILURE! - in org.shrinkwrap.springboot.app.SpringBoot13Test
should_create_valid_spring_boot_archive(org.shrinkwrap.springboot.app.SpringBoot13Test)  Time elapsed: 3.842 sec  <<< ERROR!
java.net.ConnectException: Connection refused
	at java.net.PlainSocketImpl.socketConnect(Native Method)
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
	at java.net.Socket.connect(Socket.java:589)
	at java.net.Socket.connect(Socket.java:538)
	at sun.net.NetworkClient.doConnect(NetworkClient.java:180)
	at sun.net.www.http.HttpClient.openServer(HttpClient.java:432)
	at sun.net.www.http.HttpClient.openServer(HttpClient.java:527)
	at sun.net.www.http.HttpClient.<init>(HttpClient.java:211)
	at sun.net.www.http.HttpClient.New(HttpClient.java:308)
	at sun.net.www.http.HttpClient.New(HttpClient.java:326)
	at sun.net.www.protocol.http.HttpURLConnection.getNewHttpClient(HttpURLConnection.java:1169)
	at sun.net.www.protocol.http.HttpURLConnection.plainConnect0(HttpURLConnection.java:1105)
	at sun.net.www.protocol.http.HttpURLConnection.plainConnect(HttpURLConnection.java:999)
	at sun.net.www.protocol.http.HttpURLConnection.connect(HttpURLConnection.java:933)
	at org.shrinkwrap.springboot.app.SpringBoot13Test$1.call(SpringBoot13Test.java:79)
	at org.shrinkwrap.springboot.app.SpringBoot13Test$1.call(SpringBoot13Test.java:73)
	at org.awaitility.core.AbstractHamcrestCondition$1.eval(AbstractHamcrestCondition.java:50)
	at org.awaitility.core.ConditionAwaiter$ConditionPoller.run(ConditionAwaiter.java:215)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)


Results :

Tests in error: 
  SpringBoot13Test.should_create_valid_spring_boot_archive » Connect Connection ...

```

**Comment by MatousJobanek** [comment]
The implementation LGTM.
The only thing I'm thinking about is this part: 

```
.addLibs(Maven.resolver()
        .resolve("org.springframework.boot:spring-boot-starter-web:1.3.5.RELEASE")
        .withTransitivity()
        .as(JavaArchive.class)
)

.addLauncher(Maven.resolver()
        .resolve("org.springframework.boot:spring-boot-loader:1.3.5.RELEASE")
        .withTransitivity()
        .as(JavaArchive.class))
```
this means that every time when the `SpringBoot13Archive` is being created, these dependency has to be resolved and added to the archive? 
If yes, would be good to have additional methods such as:
`addSpringBootStarter("1.3.5.RELEASE")`
`addStringBootLoader("1.3.5.RELEASE")`
that would use Resolver underneath - to avoid duplications in the test classes...?

**Comment by lordofthejars** [comment]
I will check what is happening to ftest. I think that what you mention should be done at arquillian level. I mean ShrinkWrap does not offer any upper level operation and it will be Arquillian who adds the dependencies in the appender.

**Comment by MatousJobanek** [comment]
Yeah, I know that SW doesn't offer any upper level operation. The reason why I came with the proposal is that this is not pure ShrinkWrap - it's different project with additional operation... It was just an idea...

**Comment by eddumelendez** [comment]
I have another proposal:

```
springBootVersion("1.3.5.RELEASE")
.addStarters("org.springframework.boot:spring-boot-starter-web", "org.springframework.boot:spring-boot-starter-security")
.addLibs(...)
````

* Set the version for spring-boot-starters through one method, just one place.
* Another method to provide the different starters, starters should use the version defined previously.
* `addLibs` to provide any third-party library
* In this proposal the spring-boot-loader shouldn't be defined by the developer since it's mandatory for spring-boot.

**Comment by lordofthejars** [comment]
@ALRubinger any preference?

**Comment by ALRubinger** [comment]
Hi, all.

Looks like great progress, and I've scheduled in a few hours tomorrow to review it properly.  I'll admit that I'm not up to speed with the Boot grammars you're looking to represent here, so I'll likely be following up then with some questions.

Has anyone yet set up an authoritative build job to run on https://shrinkwrap.ci.cloudbees.com/ ?  If you need admin access to add a job, please DM me your preferred account (e.g. I log in using Google).

**Comment by bartoszmajsak** [comment]
@ALRubinger take your time :)

> Has anyone yet set up an authoritative build job to run on https://shrinkwrap.ci.cloudbees.com/ ? If you need admin access to add a job, please DM me your preferred account (e.g. I log in using Google).

Any particular reason for using cloudbees for Shrinkwrap? Other alternatives we have is travis or circle-ci.

**Comment by ALRubinger** [comment]
> Any particular reason for using cloudbees for Shrinkwrap? Other alternatives we have is travis or circle-ci.

Only because that's where everything is now, and I'd like for everything to be in the same place for the ShrinkWrap org.

If we want to discuss where the org's repos *should* be, that's a conversation I'm open to having.  But I'll be advocating on a "eat own dogfood" mentality that likely points to Jenkins Pipelines on OpenShift.

**Comment by ALRubinger** [comment]
I have brought this in and pushed to shared branch on the upstream:

  https://github.com/shrinkwrap/shrinkwrap-springboot/tree/1

@lordofthejars why is this PR from a springboot-1 repo?  Separate repo where a branch would have done nicely?

Still seeing the test failure:

  ``
Running org.shrinkwrap.springboot.app.SpringBoot13Test
Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 4.353 sec <<< FAILURE! - in org.shrinkwrap.springboot.app.SpringBoot13Test
should_create_valid_spring_boot_archive(org.shrinkwrap.springboot.app.SpringBoot13Test)  Time elapsed: 4.285 sec  <<< ERROR!
java.net.ConnectException: Connection refused
	at java.net.PlainSocketImpl.socketConnect(Native Method)
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
	at java.net.Socket.connect(Socket.java:589)
	at java.net.Socket.connect(Socket.java:538)
	at sun.net.NetworkClient.doConnect(NetworkClient.java:180)
	at sun.net.www.http.HttpClient.openServer(HttpClient.java:432)
	at sun.net.www.http.HttpClient.openServer(HttpClient.java:527)
	at sun.net.www.http.HttpClient.<init>(HttpClient.java:211)
	at sun.net.www.http.HttpClient.New(HttpClient.java:308)
	at sun.net.www.http.HttpClient.New(HttpClient.java:326)
	at sun.net.www.protocol.http.HttpURLConnection.getNewHttpClient(HttpURLConnection.java:1169)
	at sun.net.www.protocol.http.HttpURLConnection.plainConnect0(HttpURLConnection.java:1105)
	at sun.net.www.protocol.http.HttpURLConnection.plainConnect(HttpURLConnection.java:999)
	at sun.net.www.protocol.http.HttpURLConnection.connect(HttpURLConnection.java:933)
	at org.shrinkwrap.springboot.app.SpringBoot13Test$1.call(SpringBoot13Test.java:79)
``

Let's review this again when we have a run on CI as proposed above.

**Comment by bartoszmajsak** [comment]
> Only because that's where everything is now, and I'd like for everything to be in the same place for the ShrinkWrap org.

Fair point. I'm only linked to ARQ builds so I wasn't aware where it sits as there is no notification in my mailbox nor on PRs.

> If we want to discuss where the org's repos should be, that's a conversation I'm open to having. But I'll be advocating on a "eat own dogfood" mentality that likely points to Jenkins Pipelines on OpenShift.

I'm totally up for "eat own dogfood". From my experience Cloudbees is really limited for OSS and we don't really have control over it.

**Comment by lordofthejars** [comment]
@ALRubinger Test was a flaky test, now it is fixed. 

About the branch. Usually in the projects I worked in Arquillian, the first commit comes from master branch, and it is after first alpha release that we start working on feature branches, fir this reason this first commit comes from master, but in next commits I will do it from feature branches.

The problem with the name spring-boot-1 is that when I forked the project I already had one project called shrinkwrap-spring-boot so github named in that way, to not starting renaming everything and screw everything, my repo is called like this. What I am going to do when PR is accepted is to remove both projects of my account and fork again.

**Comment by bartoszmajsak** [comment]
Do we still have any open points here or can we **squash & merge** it?

**Comment by lordofthejars** [comment]
?¿ We merge it?

**Pull request merged**